content/blog/using-apache-spark-dataframes-for-processing-of-tabular-data.md

```diff
@@ -0,0 +1,362 @@
+
+
+<font color="green">// How many categories are there</font>
+sqlContext.sql("SELECT distinct Category FROM sfpd").collect().foreach(println)
+
+
+<font color="#005CB9">// [ASSAULT]
+// [MISSING PERSON]
+// [TREA] . . .</font>
+</pre>
+
+<pre>
+<font color="green">// What are the top 10 Resolutions ?</font>
+sqlContext.sql("SELECT Resolution , count(Resolution) as rescount FROM sfpd group by Resolution order by rescount desc limit 10").collect().foreach(println)
+<font color="#005CB9">// [NONE,1063775]
+// [ARREST, BOOKED,414219]
+// [ARREST, CITED,154033] . . .</font>
+</pre>
+
+<pre>
+<font color="green">// What are the top 10 most incident Categories?</font>
+val t =  sqlContext.sql("SELECT Category , count(Category) as catcount FROM sfpd group by Category order by catcount desc limit 10")
+
+
+t.show()
+<font color="#005CB9">// Category       catcount
+// LARCENY/THEFT  353793
+// OTHER OFFENSES 253627
+// NON-CRIMINAL   186272\. . .</font>
+
+
+<font color="green">// The results of SQL queries are DataFrames and support RDD operations.
+// The columns of a row in the result can be accessed by ordinal</font>
+t.map(t => "column 0: " + t(0)).collect().foreach(println)
+<font color="#005CB9">// column 0: LARCENY/THEFT
+// column 0: OTHER OFFENSES
+// column 0: NON-CRIMINAL
+// column 0: ASSAULT …</font>
+</pre>
+
+## The physical plan for DataFrames
+
+The <a target='\_blank'  href='https://databricks.com/blog/2015/04/13/deep-dive-into-spark-sqls-catalyst-optimizer.html'>Catalyst query optimizer creates the physical Execution Plan</a> for DataFrames as shown in the diagram below:
+
+![](/img/blog_sparkdataframes_image3.png)
+
+**Print the physical plan to the console**
+
+DataFrames are designed to take the SQL queries constructed against them and optimize the execution as sequences of Spark Jobs as required. You can print the physical plan for a DataFrame using the explain operation as shown below:
+
+<pre>
+<font color="green">//  Prints the physical plan to the console for debugging purpose</font>
+auction.select("auctionid").distinct.**explain**()
+
+
+<font color="#005CB9">// == Physical Plan ==
+// Distinct false
+// Exchange (HashPartitioning [auctionid#0], 200)
+//  Distinct true
+//   Project [auctionid#0]
+ //   PhysicalRDD   //[auctionid#0,bid#1,bidtime#2,bidder#3,bidderrate#4,openbid#5,price#6,item#7,daystolive#8], MapPartitionsRDD[11] at mapPartitions at ExistingRDD.scala:37</font>
+</pre>
+
+## Summary
+
+In this blog post, you’ve learned how to load data into Spark DataFrames, and explore data with Spark SQL. If you have any further questions, or want to share how you are using Spark DataFrames, please add your comments in the section below.
+
+**Want to learn more?**
+
+* <a target='\_blank'  href='https://spark.apache.org/docs/1.3.0/sql-programming-guide.html#dataframes'>Spark SQL and DataFrame Guide</a>
+* <a target='\_blank'  href='http://shop.oreilly.com/product/0636920028512.do'>Learning Spark - O'Reilly Book</a>
```

Comment:
Use the Markdown below to list items.
Also, you may want to use the link of "Learning Spark, 2nd edition (date 2020)" rather than the link to the 1st edition (date: 2015).

**Want to learn more?**

* [Spark SQL and DataFrame Guide](https://spark.apache.org/docs/1.3.0/sql-programming-guide.html#dataframes)

* [Learning Spark - O'Reilly Book](https://www.oreilly.com/library/view/learning-spark-2nd/9781492050032/)

Reply:
done